### ISSUE TITLE
Battery integration

### ISSUE THREAD

**Issue by sragas** [WillCodeForCats/solaredge-modbus-multi#22]
Are there any plans to integrate connected batteries?

**Comment by WillCodeForCats** [comment]
Yeah eventually. Right now I'm focusing on getting another integration released (my tekmar-482 integration) and once that's out I'll revisit this one. I'm planning to rewrite this integration since I'm finding how it was written is limiting to things I want to do with it, and he structure I used for the tekmar-482 integration is more flexible for me.

**Comment by skl283** [comment]
Will also appreciate the battery integration to use your Integration! :-)

**Comment by WillCodeForCats** [comment]
The [rewrite-for-2.0](https://github.com/WillCodeForCats/solaredge-modbus-multi/tree/rewrite-for-2.0) branch is at a point where it should detect batteries and create device entries for them, but there aren't any sensor entities yet.

**Comment by WillCodeForCats** [comment]
Anyone who has batteries please try the [rewrite-for-2.0](https://github.com/WillCodeForCats/solaredge-modbus-multi/tree/rewrite-for-2.0) branch and report if it detects your batteries or not. It will have some basic diagnostic sensors and show a device for each battery (up to 2 per inverter for up to 32 inverters).

Because battery registers are not technically officially supported by SolarEdge, auto-detection for batteries is disabled by default. To enabled battery detection, go to the device options and enable it.

<img width="364" alt="Screen Shot 2022-06-02 at 11 23 09 AM" src="https://user-images.githubusercontent.com/48533968/171700220-bffa6088-da9c-4869-85a2-d11e87638be3.png">

<img width="443" alt="Screen Shot 2022-06-02 at 11 22 30 AM" src="https://user-images.githubusercontent.com/48533968/171700246-86fc1e17-0a2c-4fdc-adb8-7b40b8cc02a1.png">

**Comment by Doonga** [comment]
Hi, I gave this a shot, I got devices for the inverter and the 2 meters, but no batteries. I have the newer SolarEdge Battery Banks BAT-10K1P, there are two of them.

I rolled back to the normal version for now since I wasn't sure if all of the device names changed. I'll need to go edit a bunch of stuff to keep things working.

**Comment by WillCodeForCats** [comment]
Inverter entity names will be the same, meter entities will be slightly different since auto detect will scan every inverter for all 3 meters, which means it's possible to have more than 3 meters, so it names the meter entities with the inverter ID and the meter ID instead of fixed numbers.

If you set debug logging it will log something for batteries. Basically the way auto detect works is it tries to read some basic info registers and if the read is OK then it adds the device. If the read returns any kind of error it ignores that location. The battery registers have been taken from an engineering pdf for the LG batteries where only 2 batteries per inverter were supported.

To enable debug logging:
```
logger:
    default: warning
    logs:
        custom_components.solaredge_modbus_multi: debug
```

And then in my logs the important lines are:

```
DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 1: Exception Response(131, 3, IllegalAddress)
DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 2: Exception Response(131, 3, IllegalAddress)
```

Which means there aren't any batteries at the battery register, in my case because I don't have any so the address doesn't exist. If you see those exceptions and you're not using the old LG batteries, then that means it's likely SolarEdge changed the battery registers or isn't using them.

**Comment by WillCodeForCats** [comment]
I just pushed some updates to logging attributions so you'll need to re-download.

**Comment by Doonga** [comment]
Hmm yep looks like the new ones aren't using the registers... That's very unfortunate...

`2022-06-02 19:18:17 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 1: Exception Response(131, 3, IllegalAddress)
2022-06-02 19:18:17 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 2: Exception Response(131, 3, IllegalAddress)
`

**Comment by WillCodeForCats** [comment]
Well, I'll keep an eye out for any information about registers for the new Energy Bank system. Otherwise it's anyone's guess if they moved to different addresses or what.

**Comment by Doonga** [comment]
Yep, no worries. Hopefully somewhere down the road. Thanks!

**Comment by WillCodeForCats** [comment]
Pre-release v2.0.0-pre.1 is available now.

**Comment by muenchris** [comment]
I have two Inverter and each inverter has one battery attached. But I do not think SolarEdge is exposing the batteries via Modbus TCP at all. Using modbus test tools, there are no values exported where they should be according to the SunSpec. I guess the inverters are collecting the battery info via RS485 but do not copy the values into the Modbus TCP range. Luckily the LG batteries have a mini WebServer on board that can easily be used to read all BMS values

**Comment by WillCodeForCats** [comment]
LG batteries should be detected. My code may have been off by one count on the battery info read block, I will release a pre.7 with what I think might be the correct value.

**Comment by muenchris** [comment]
Nice, I can test right away if you like

**Comment by WillCodeForCats** [comment]
[Release v2.0.0-pre.7](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.0-pre.7)

**Comment by WillCodeForCats** [comment]
If you're using HACS you will have to click "update information" on the integration card in HACS.

**Comment by muenchris** [comment]
I have auto-detect batteries on and restarted HA. After the restart - will it try to auto-detect the batteries automatically or do I have to do something? I updated the HACS integration to Pre7 first then restarted. The Integration still shows 5 devices (2 inverters + 3 Meters) - no batteries.

Also, do the batteries require specific DeviceIDs?

**Comment by WillCodeForCats** [comment]
You have to enable auto battery detect in config options:

https://github.com/WillCodeForCats/solaredge-modbus-multi/wiki/Configuration#integration-configuration-options

It's disabled by default.

**Comment by muenchris** [comment]
yes I had it on before the restart

**Comment by WillCodeForCats** [comment]
Hmm, well, I don't know then. I don't have batteries so anything I try for them is random guessing based on other information I can find.

**Comment by muenchris** [comment]
As I wrote before, I used a Modbus tool to see if the SolarEdge inverter somehow exposes the batteries, but there is no data returned where the SunSpec is referencing it. I guess that SE is not exposing it via Modbus TCP.
My next step is to see if I can somehow use the same path the SE SetApp is going. That seams to have much more information including the optimizers of each panel. The only tricky thing with that is that the server is only on during the "Commissioning" (with a Phone/PC connected with Wifi). Also they are using Protobuf  and that is hard to decode without the .proto file.

**Comment by WillCodeForCats** [comment]
They're supposed to start at 57600 for batteries, they do not use sunspec storage model.

**Comment by muenchris** [comment]
I see, makes sense. I will try to see what I get at that address

**Comment by WillCodeForCats** [comment]
Thanks for trying. In a perfect world, SolarEdge would expand the modbus interface to support the sunspec storage model.

**Comment by muenchris** [comment]
yes, I hope they do someday

**Comment by muenchris** [comment]
I just had my Span panels commissioned with my two SolarEdge inverters and they told me that SolarEdge has to "Enable the battery" for the Modbus interface. Looks like SolarEdge is limiting local access to the system more and more. 

So if anyone does not see the battery, its because SolarEdge has not enabled the modbus TCP interface for the batteries.

**Comment by WillCodeForCats** [comment]
Is that something they will do if you request it?

**Comment by muenchris** [comment]
I am not sure. Span is doing this for me but I am not sure anyone can request it. I will try to find out

**Comment by mbillings6** [comment]
I just downloaded pre7 to test it, I have a single inverter with the LG batteries. I have been searching for information on this an I have seen some that say they are using references up in the 62786 range and above for batteries lately.

**Comment by WillCodeForCats** [comment]
FYI pre.7 is the same as release v2.0.0, just without the pre-release flag.

If the registers changed I wonder if it's related to firmware updates to support new their new batteries.

**Comment by mbillings6** [comment]
I am not sure how helpful this will be but I found this document related to SunSpec on SolarEdge power control protocols. I am still pretty new at working with Modbus TCP or Modbus period so I have yet to make sense of it.

https://www.photovoltaikforum.com/core/attachment/88445-power-control-open-protocol-for-solaredge-inverters-pdf/

**Comment by WillCodeForCats** [comment]
That PDF is what gives the 57600 starting register for batteries, but it's also 5 years old.

What I will have to try myself (when I have some free time) is simulate a modbus slave with data to make sure the integration at least works with a simulated battery since I don't have a real one.

**Comment by mbillings6** [comment]
If you want, we try it on my setup. I have the SE3800H-USS3BBC14 inverter and the LG RESU10H Prime for the battery. I have tried accessing registers above like 41000 and it sends back and illegal address message.

**Comment by Dave-Casey** [comment]
I'm also using the LG RESU10H Prime battery attached to my LG RESU10H Prime.  I have enabled `Auto-detect batteries` but do not see it.  Let me know how I can help.

**Comment by WillCodeForCats** [comment]
See these instructions for debug logging and what to look for:

https://github.com/WillCodeForCats/solaredge-modbus-multi/issues/22#issuecomment-1145413524

**Comment by muenchris** [comment]
My Span.io panels finally work with my two inverters and the two batteries. There was some "Magic" going on between the SolarEdge Support and the Span.IO developers leading up to SolarEdge "enabling the registers" for my inverter. After a restart of my inverter (something they did without telling me and occurred right in the middle of a Zoom Call), the Span App now shows the correct battery status. Since they are accessing my inverter via Modbus TCP, I assume that "Magic" is something SolarEdge needs to enable for the batteries to show up. 

Unfortunately Span is keeping the Modbus TCP connection open to the inverter and since the SolarEdge inverter is a Highlander (there can be only one...connection), I cannot access it with the this HA integration as well. 

One way of testing would be to disconnect the Span panel temporarily and then try to use this integration. My fear is that after that, my span app might not work anymore. 

My best advise to you all: Contact SolarEdge Support and ask them to enable the "full" Modbus TCP registers.

**Comment by WillCodeForCats** [comment]
There's a modbus proxy add-on thing to deal with the there can only be one issue. I don't remember the name offhand though. Maybe this one? https://github.com/Akulatraxas/ha-modbusproxy/ 

Since this integration closes the connection after reading technically it can coexist with other stuff, but only as long as those other things also close the connection between their polling cycles.

**Comment by muenchris** [comment]
yes, my software and your integration does that...but Span does not. And since they discover the invertersomehow, I am not sure if it would work with their implementation. 
All these integrations are still very uncoordinated. I hope things are getting better over time

**Comment by mbillings6** [comment]
I was on the phone with Solar Edge Support earlier, and I was able to confirm the Nov 2017 document is the latest on the Power Control Open Protocol. However, they were not really able to help me outside of that. Based on what @muenchris mentioned previously about Solar Edge doing something to unlock the addition registers and my conversation and research. I would say the answer is probably in the Power Control section of the inverter. I may eventually test this theory.

**Comment by WillCodeForCats** [comment]
I suppose if I had a battery, the first thing I would try is checking the value of `0xF142` and try setting `0xF142 = 1` (if it was `0`) then commit it with `0xF100 = 1`.

**Comment by rzulian** [comment]
I do have a LG battery connected to the inverter 1, that is not being recognised by this custom component, but it works with others. I'll check what's wrong

**Comment by TheGardenMonkey** [comment]
Which other approach has worked for you? I have an LG as well, but using the solaredge_modbus python module directly I am not able to get the batteries (also don't seem to get accurate values, but no battery). 

> I do have a LG battery connected to the inverter 1, that is not being recognised by this custom component, but it works with others. I'll check what's wrong

Can you point me to some instructions on how to try this out?
> I suppose if I had a battery, the first thing I would try is checking the value of `0xF142` and try setting `0xF142 = 1` (if it was `0`) then commit it with `0xF100 = 1`.

**Comment by rzulian** [comment]
> 

I've been working on this fork https://github.com/rzulian/home-assistant-solaredge-modbus/tree/mpredfearn/multiple_inverters
I've enabled multiple inverters. This is working just fine, but it is not maintained so I was trying this component instead.

**Comment by rzulian** [comment]
I'm using v2.0.3, accordingly to the manifest.json
Thisn is nthe line that I've added after the try.
```
        if battery_info.isError():
            _LOGGER.debug(f"Inverter {self.inverter_unit_id} battery {self.battery_id}: {battery_info}")
            raise RuntimeError(battery_info)
        _LOGGER.debug(f"Inverter {self.inverter_unit_id} battery {self.battery_id} found: {battery_info}")
```

**Comment by WillCodeForCats** [comment]
Oh, so that was a line you added. It didn't make sense to me since I didn't see it in my copy.

It's probably just an issue with the decoded_common block then, possibly I have a type wrong or the register count is off and it's trying to decode off the end.

**Comment by WillCodeForCats** [comment]
I see it now, it's a simple mistake.

On line 711 `decoded_common` should be `self.decoded_common`

It's probably bailing out there.

**Comment by WillCodeForCats** [comment]
I pushed the correction to the `main` branch.

**Comment by TheGardenMonkey** [comment]
Still no battery detected for me after using the correct code. For reference these are the devices I have:

SolarEdge SE7600H-USS3BBC14 
SolarEdge SE-RGMTR-1D-240C-B 
SolarEdge SE-MTR-3Y-240V-A 
LG RESU10H

It feels like something else must be blocked since I am not able to see the battery through the python clients directly.

**Comment by rzulian** [comment]
I've a working version which is loading the battery. I cannot create a pull request, because my formatting is adding a lot o changes. @WillCodeForCats you can check my changes in my fork, and if you want to share your formatting settings, I can create a pull request.

**Comment by WillCodeForCats** [comment]
I don't have any formatting settings, I use BBEdit on MacOS to do it all manually.

Can't you just change what the issue without making formatting changes? If it's the variable I already pushed just try that.

**Comment by WillCodeForCats** [comment]
I used a modbus simulator to find and fix other errors related to the battery device class. I will publish a pre-release shortly.

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.1](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.1)

**Comment by rzulian** [comment]
I've tested it and it works. Are you going to create the sensors for the batteries?

**Comment by WillCodeForCats** [comment]
Yes, that's the plan. There should be one sensor already: the device sensor.

Can you provide all the attributes for each device sensor it created?

**Comment by TheGardenMonkey** [comment]
I now see the following with debug enabled and the new beta:

```
2022-07-13 16:40:13 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 1: Exception Response(131, 3, IllegalAddress)
2022-07-13 16:40:13 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 2: Exception Response(131, 3, IllegalAddress)
```

Are there some suggestions for how i can try to figure out what are the proper registers for my configuration?

**Comment by rzulian** [comment]
> Yes, that's the plan. There should be one sensor already: the device sensor.
> 
> Can you provide all the attributes for each device sensor it created?



> Yes, that's the plan. There should be one sensor already: the device sensor.
> 
> Can you provide all the attributes for each device sensor it created?

these?

<img width="319" alt="Screenshot 2022-07-13 at 23 23 51" src="https://user-images.githubusercontent.com/29941924/178839104-8ba1d9c0-5bb0-4b56-b0bb-d90c6c3de964.png">

**Comment by WillCodeForCats** [comment]
Yeah, but they don't look quite right. It says you have a 2 watt battery.

Can you try changing line 704 to `battery_info.registers, byteorder=Endian.Big, wordorder=Endian.Little` and see how that changes the reported values?

**Comment by WillCodeForCats** [comment]
Change these lines: c2da772e079c6f2059dde5b288e3405e9dd3d2d0

Or download the branch: https://github.com/WillCodeForCats/solaredge-modbus-multi/archive/refs/heads/22-battery-integration.zip

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.2](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.2)

**Comment by WillCodeForCats** [comment]
> 
> Are there some suggestions for how i can try to figure out what are the proper registers for my configuration?

All inverters should use the same registers. These messages mean your inverter says they are invalid. It's been suggested that some inverters (newer systems?) have them disabled and requireSolarEdge to enable them, but since I don't have batteries personally I don't have any experience with that so maybe someone else will have a suggestion.

**Comment by rzulian** [comment]
> Release v2.0.3-battery.2

the different wordorder is working properly

**Comment by WillCodeForCats** [comment]
Can you post the device attributes with the changed wordorder?

**Comment by rzulian** [comment]
<img width="313" alt="Screenshot 2022-07-14 at 08 34 52" src="https://user-images.githubusercontent.com/29941924/178916694-4f9413f5-5c99-4b79-9976-64c512f5d080.png">

**Comment by WillCodeForCats** [comment]
Commit 87199f6d2d0a5ddc85640a24c21e5cea2308194f should clean up the manufacturer and model.

**Comment by TheGardenMonkey** [comment]
> > Are there some suggestions for how i can try to figure out what are the proper registers for my configuration?
> 
> All inverters should use the same registers. These messages mean your inverter says they are invalid. It's been suggested that some inverters (newer systems?) have them disabled and requireSolarEdge to enable them, but since I don't have batteries personally I don't have any experience with that so maybe someone else will have a suggestion.

Got it and thank you, I will try to reach out to SolarEdge directly...they seem to not want to deal with home owners directly and my installer has no clue what I am looking to do so I may have to give up on this and stick with the cloud api, but I will report back if I get an answer in case it helps someone else.

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.4](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.4)

This pre-release adds sensors for batteries.

**Comment by WillCodeForCats** [comment]
Energy sensors are broken, but commit db0d611aea827835da31f3448656b8b17f51255e should fix that.

**Comment by WillCodeForCats** [comment]
I expect max temperature is wrong, but I don't know what range to validate it against.

State of Energy looks wrong to me since it's supposed to be Battery Available Energy / Battery Max Energy, so it should be 9.17/9.8 = 93%

**Comment by WillCodeForCats** [comment]
I suppose I could just make my own sensor for Available/Max=Charge% instead of using the value from the register.

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.5](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.5)

**Comment by youtoofan** [comment]
My battery SOC says 100% in Home Assistant, while the official state is 82%. Also max temperature seems a bit off:

![image](https://user-images.githubusercontent.com/1804461/179392464-757467b8-45de-4b8c-81c2-906022bda852.png)
![image](https://user-images.githubusercontent.com/1804461/179392473-72847949-367d-4b39-ae85-0d2d095255c0.png)

**Comment by WillCodeForCats** [comment]
What does your State of Energy (SOE) say? State of Energy and State of Health are reported values from the device.

State of Charge is not reported, it is based on calculated values where available/maximum is 9623/9623=100% which is correct.

Maximum temperature I will probably just change to default disabled since that is the reported value but obviously not a realistic value.

**Comment by WillCodeForCats** [comment]
Commit 21812d22f06a70ad6cc95946a2dbe322e012fbba makes the battery max temperate sensor disabled by default.

**Comment by WillCodeForCats** [comment]
Oh, wait, I see what's going on. The DC voltage was mixed up as the max temp. I don't know how I missed that. I will do another pre-release to fix that.

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.6](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.6)

**Comment by youtoofan** [comment]
Max temp now says 0.0 C°, and SOE shows indeed the correct value.

**Comment by WillCodeForCats** [comment]
Other sample data I have found for LG batteries also shows 0 for max temp.

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.7](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.7)

This will probably be the final pre-release unless something else needs fixing.

**Comment by phpdiddy** [comment]
Just as another data point, I have two SolarEdge SE7600H inverters and two LG RESU 16H batteries. I cannot see battery information from this integration. Log shows:

`2022-07-18 19:02:14 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 1: Exception Response(131, 3, IllegalAddress)
2022-07-18 19:02:14 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 1 battery 2: Exception Response(131, 3, IllegalAddress)
2022-07-18 19:02:15 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 2 battery 1: Exception Response(131, 3, IllegalAddress)
2022-07-18 19:02:16 DEBUG (MainThread) [custom_components.solaredge_modbus_multi.hub] Inverter 2 battery 2: Exception Response(131, 3, IllegalAddress)`

Via SolarEdge SetApp, I can see that the batteries are connected and working as expected. Each inverter has a single battery connected. One thing I did notice is that the MODBUS device IDs for the batteries are not sequential. I know that caused me some issues initially in getting both of my inverters showing up, but not sure if it matters for batteries.

**Comment by WillCodeForCats** [comment]
Only the inverter device ID matters to the integration. Meters and batteries are "hidden" behind the inverter, which basically acts as a proxy for that data. The order of modbus ID for meters and batters should only affect the order in which they are detected by the inverter.

The "IllegalAddress" response is the inverter saying "your request does not exist here".

**Comment by phpdiddy** [comment]
That's what I figured, but wanted to be sure. So it seems either SolarEdge or LG are not delivering this information via modbus on some installations. If anyone has any insight into what settings may need to be adjusted for this to work, I'd be happy to try them. SetApp does show the right information on the battery manufacturer and size, so I think it's more likely the inverters just aren't exposing it.

**Comment by muenchris** [comment]
Yes, something must have change (by SolarEdge) in my inverter. My Span.io panels can now read all the information from the inverter including the batteries. Unfortunately Span is keeping the Modbus connection open all the time and that prevents other systems from accessing the SolarEdge modbus interface (highlander interface: "there can be only one...connected". 
Some time in the future I plan on disconnecting the Span panel from the inverter and access the inverter with Seth's connector to see if this works. Right now I am testing the Span system to see if it can provide me with the data I want (their interface is a standard JSON REST service and much more modern than this old Modbus crap)

**Comment by gferen** [comment]
How have anyone managed to display/discover LG resu battery?  i was wondering, since  different hacs SolarEdge Modbus integration shows battery just ok. as well as info from official integration with API from SolarEdge?

**Comment by muenchris** [comment]
I tested this integration with my second inverter and that is now showing the battery. So I assume whatever SolarEdge "enabled" did work on my system. How is your LG battery connected to your SolarEdge inverter? Most likely with CAN or RS485. The inverter uses some internal (not published) way to read the data and sends it to their cloud. The Modbus interface we are all using is mapping the CAN/RS485 data from the LG Battery into a certain address range. But from what I learned, that is not always enabled on the inverter. Span needed that and they asked SolarEdge to enable it for my system. How this can be done by an end-user is not clear.

**Comment by WillCodeForCats** [comment]
> How have anyone managed to display/discover LG resu battery? i was wondering, since different hacs SolarEdge Modbus integration shows battery just ok. as well as info from official integration with API from SolarEdge?

Yes several people have reported it works.

**Comment by muenchris** [comment]
Mine only worked after SolarEdge "enabled" something (that they did not tell me)

**Comment by gferen** [comment]
yes it is RS485 bus. meter and battery are connected to rs485-2 to inverter. they communicate well.  I have access to battery's BMS   config UI, I have "admin - installer" access to inverter. (not just locally to web page 172.0.16.xxx) but with setApp, as well to monitoring platform. i asked my installer to make me installer profile, so i manage whole system my self. Unless in situation Span-SE to enable this it is something really hidden, even from installers

**Comment by muenchris** [comment]
is your webpage "unlocked" or do you have to connect with Wifi first?
Is your LG battery also connected to your network via LAN? (If not, you could ask your installer to do this. There is a webserver in the LG battery giving more information than the Modbus ever does)

**Comment by gferen** [comment]
battery is connected to inverter via rs485-2 (second modbus line on inverter). 
battery is connected to wifi. i access BMS wia web-UI, and use default credentials to login.
I ask my installer only to crimp new connectors on solar cables when I add new panes. all the wiring i do alone.
the web-ui has tons of information, configurations, administration, etc.

**Comment by gferen** [comment]
@muenchris not sure what you mean unlocked.... "is your webpage "unlocked" or do you have to connect with Wifi first?"

what i did, i checked for weird WIFI SSIDs next to inverter and battery. I found 1 persistent that I connected with password written inside of battery. After that I set Static IP on wifi settings, and left SoftAP alone. and like that i am accessing BMS via local IP over Wifi. As well i can still connect to softAP. I belive softAP is used when you have mobile RESU app and then smart app connects to BMS. Just like setApp connects to inverter with scanning QR code

**Comment by muenchris** [comment]
yes....and the most funny thing is that the security does not include the sub-pages. So if you access the 'page" http://<yourLGIP>/getbmsinfo you get just the data back without any auth. Since its a HTML block, you need to parse it yourself but that is very easy.

**Comment by muenchris** [comment]
@gferen I have to push the little doggle on the inverter to enable the SoftAP. And if I dont refresh the website for more then 15 seconds, the SoftAP disconnects again. "Unlocked" for me means that you dont have to push the little toggle (basically the SoftAP is always on).
My inverters are all connected via Ethernet. And that connection uses a different IP than the SetApp is using

**Comment by gferen** [comment]
This is what i Access in lg battery.
![BMS](https://user-images.githubusercontent.com/1295008/180038678-1b37c4ec-caa0-441d-9785-352190defff5.png)


I got i now for unlocked etc.  and yes for local access of commissioning page there is 172.x.x.x As far as i know even if inverter is connected to LAN you cant get to it by LAN IP.

**Comment by muenchris** [comment]
yes, and the SoftIP is only on, if I touch the toggle as it says in the SetApp

**Comment by WillCodeForCats** [comment]
If anyone can verify the status sensor is accurate, such as charging, discharging, etc. then I think we can merge the PR and do a release.

**Comment by muenchris** [comment]
my SOC shows 100% although its 72.85 according to SolarEdge and the LG battery website. Maximum energy is correct but "Available Energy" should be something around 7000.
In the diagnostics section I see "State Of Energy" 74% <- that is actually the SoC.
Voltage and Current and DC Power looks about right.

**Comment by WillCodeForCats** [comment]
Yeah I can drop the calculated state of charge sensor if it's too confusing.

**Comment by muenchris** [comment]
Does it not show in the Modbus registers?

**Comment by WillCodeForCats** [comment]
No, there is no SOC register, only SOE and SOH.

**Comment by WillCodeForCats** [comment]
I'll just remove the state of change sensor. It can be duplicated with a sensor template if someone really wants the available/max=% value as a sensor.

**Comment by muenchris** [comment]
just the available is not showing the correct value. on my battery it shows the same amount as max.

**Comment by WillCodeForCats** [comment]
The "Available Energy" sensor is a register value from the inverter.

If you turn on debug logging look for the hex values of `B_Energy_Max` and `B_Energy_Available`. The raw data should be in watts and then they are converted to kW in the sensor.

**Comment by muenchris** [comment]
(how do I turn debug on :) ?)

**Comment by WillCodeForCats** [comment]
```
logger:
    default: warning
    logs:
        custom_components.solaredge_modbus_multi: debug
```

**Comment by muenchris** [comment]
I am running HASS in a VM. What file is that in?

**Comment by WillCodeForCats** [comment]
File `/config/configuration.yaml` then restart.

**Comment by muenchris** [comment]
restarted. Thx. Where do I find the raw values?

**Comment by muenchris** [comment]
looks like I have another problem with HA. My VM says "failed to save stream data/journal/..." No Space left on device.
the VM HD is 127GB ...I wonder if I can turn off this journal

**Comment by thewizard-de** [comment]
Hi, i´m on the latest beta .7 and it find the LG battery but only energy export is not monitored. 
[solaredge.txt](https://github.com/WillCodeForCats/solaredge-modbus-multi/files/9175444/solaredge.txt)
![grafik](https://user-images.githubusercontent.com/10088048/180640111-3ba6d082-672b-4b97-8c0e-da38f6b11539.png)

**Comment by WillCodeForCats** [comment]
> Hi, i´m on the latest beta .7 and it find the LG battery but only energy export is not monitored.

Fixed in commit b3af09ca5c9f5c7413683326973c1965aa15cc48

**Comment by WillCodeForCats** [comment]
[Release v2.0.3-battery.8](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.0.3-battery.8)

**Comment by WillCodeForCats** [comment]
> looks like I have another problem with HA. My VM says "failed to save stream data/journal/..." No Space left on device.
> the VM HD is 127GB ...I wonder if I can turn off this journal

Not familiar with that issue. All my test and development systems are virtual machines in VMware Fusion, and my "live" home assistant install is a KVM VM.

**Comment by muenchris** [comment]
Oh sorry, I wrote that a bit misleading. It has nothing to do with your code. Its some "misbehaving" my HA is doing and then all sort of issues show up that have nothing to do with plugins or integrations. It looks like nobody has a good solution for that but its known in the community.

**Comment by WillCodeForCats** [comment]
Last call for any pre-release testing or issues before I merge PR #58 for a release.

**Comment by muenchris** [comment]
Looks good to me. "DC Power" is always zero on my battery. Is this normal?

**Comment by WillCodeForCats** [comment]
> Looks good to me. "DC Power" is always zero on my battery. Is this normal?

I wouldn't think so, but since I don't have a battery I just have to trust that what register values are is what is accurate.

**Comment by WillCodeForCats** [comment]
Merged PR #58 into main.

**Comment by WillCodeForCats** [comment]
I am converting this issue to a discussion since it still contains valuable information about battery support.
